M1FExplained/Chapter10/Exercises/Exercise02.lean

```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
+in the actual question statment.
 -/
 
-lemma part_a (a b d : ℕ) (hd : d = Nat.gcd a b) : ∃ (s t : ℕ),  d = s * a - t * b := by sorry
+lemma Int.sub_ediv_of_dvd_right {a b c : ℤ} (H : c ∣ b) : (a - b) / c = a / c - b / c := by
+  simp only [sub_eq_add_neg, Int.add_ediv_of_dvd_right <| Int.dvd_neg.2 H, Int.neg_ediv_of_dvd H]
+
+lemma Int.sub_lt_div_mul_self {a b : ℤ} (hb : 0 < b) : a - b < a / b * b := by
+  by_contra h
+  push_neg at h
+  have := Int.le_ediv_of_mul_le hb h
+  rw [Int.sub_ediv_of_dvd_right <| dvd_rfl, Int.ediv_self hb.ne'] at this
+  linarith
+
+/- 
+helper_1 says that given s * a + t * b = n, we can find new coefficients s', t' for a, b 
+with 0 < s' so the same equality holds.
+-/
+
+lemma helper_1 (a b n : ℤ) (hb : 0 < b) (hab : ∃ (s t : ℤ), s * a + t * b = n) : 
+    ∃ (s' t' : ℤ), 0 < s' ∧ s' * a + t' * b = n := by 
+  rcases hab with ⟨s, t, hst⟩
+  set p := (b - s) / b with hp
+  use s + p * b, t - p * a
+  constructor
+  · rw [hp, ← neg_lt_iff_pos_add']
+    calc
+      -s = (b - s) - b := by norm_num
+      _ < (b - s) / b * b := by exact Int.sub_lt_div_mul_self hb
+  · linear_combination hst
 
 /-
-Part b asks us to find such s and t for the examples in exercise 01, Lean provides us with these: 
+helper_2 builds on helper_1 but now adds the restriction that the new coefficient t' satisfies t' < 0.
+This effectivley solves the question, save for some changes of sign and the assumption that we can 
+initially find any s, t such that s * a + t * b = gcd(a, b).
+-/
+
+lemma helper_2 (a b n : ℤ) (ha : 0 < a) (hb : 0 < b) (hab : ∃ (s t : ℤ), s * a + t * b = n) : 
+    ∃ (s' t' : ℤ), 0 < s' ∧ t' < 0 ∧ s' * a + t' * b = n := by 
+    match helper_1 a b n hb hab with
+    |⟨s, t, ⟨h1, h2⟩⟩ => 
+    apply Or.elim (Classical.em (t < 0))
+    <;> intro ht
+    · use s, t
+      repeat (any_goals constructor)
+      all_goals assumption
+    · push_neg at ht
+      set p := (t + a) / a with hp
+      have hpos : 0 ≤ p := by 
+        · rw [hp]
+          refine Iff.mpr (Int.le_ediv_iff_mul_le ha) ?_
+          norm_num
+          exact Int.add_nonneg ht (show 0 ≤ a by exact Int.nonneg_of_pos ha)
+      use (s + p * b), (t - p * a)
+      repeat (any_goals constructor)
+      · exact add_pos_of_pos_of_nonneg h1 (show _ by exact Iff.mpr (zero_le_mul_right hb) hpos)
+      · rw [sub_neg, hp]
+        have := @Int.sub_lt_div_mul_self (t + a) a ha
+        simp at this
+        assumption
+      · linear_combination h2 
+
+/-
+The final statment of the question consistient with that in the book. helper_2 states
+the equality s' * a + t' * b = n with 0 < s' and t' < 0. 
+
+However, we want s' * a - t' * b = n with 0 < s' and 0 < t'. These are of course equivalent,
+and part_a deals with this. 
+-/
+
+lemma part_a (a b : ℤ) (ha : 0 < a) (hb : 0 < b) : 
+  ∃ (s t : ℤ), 0 < s ∧ 0 < t ∧ Int.gcd a b = s * a - t * b := by 
+  have := helper_2 a b (Int.gcd a b) ha hb (show _ by 
+    use Int.gcdA a b, Int.gcdB a b
+    simp [mul_comm]
```

Comment:
This is a non-terminal `simp`. Just close the goal with `simp [mul_comm, Int.gcd_eq_gcd_ab])`